### ISSUE TITLE
Feature Request: Provide option to store images/containers/volumes on other drives

### ISSUE THREAD

**Issue by SetTrend** [docker/for-win#11357]
My system disk (C:) is full, just filled with containers an images. My D: drive however, is ...
1. empty
2. larger
3. expandable

But there's no viable way of defining a destination folder for images, containers a/o volumes.

**So, I propose the following feature request:**

1. Provide an Data Path option in the installer allowing users to define a default folder for storing images, containers and volumes.

2. Add a new command to the Dockerfile syntax:
   `TO` {destination path}
   
3. Add a new parameter to `docker commit`, `docker container create`, `docker create`, `docker run`, and `docker volume create`:
   `--data-path` {destination path}

The destination path is supposed to be a directory wherein the corresponding image, container or volume will create its own, internal, private sub directory to store its data. So, if the destination directory was `D:\DockerData\tmp`, then an image, for example, may first create a sub directory, like `D:\DockerData\tmp\03624fa1359\` to then store its data therein.

<hr/>

Examples:

```dockerfile
FROM mcr.microsoft.com/windows/nanoserver
TO D:\DockerData\nano\
RUN echo Done
```

```batch
> docker run --name test --data-path D:\DockerData\nano\ fromImage
```

```batch
> docker volume create --data-path D:\DockerData\nano\ myStorage
```

**Comment by SetTrend** [comment]
Keep alive ...

**Comment by stephen-turner** [comment]
I'm not sure of the exact command, but I think this should be configurable in WSL 2, not in Docker. We just use the WSL 2 filesystem, WSL owns the location.

**Comment by SetTrend** [comment]
Coincidentally, I created a sibling issue at the WSL repository:
[/microsoft/WSL/issues/7128](/microsoft/WSL/issues/7128).

Unfortunately, they prematurely closed it 😞

**Comment by stephen-turner** [comment]
We have regular meetings with the WSL engineers, we'll discuss this at our next meeting.

**Comment by SetTrend** [comment]
Adding additional administrative tools to support the suggested feature would be nice, e.g.:
<br/>

#### Moving an image to a new location:

```powershell
> docker image mv myImage:latest E:\DockerData\nano\
```
(This would basically be a copy/redirect/delete operation.)
<br/>

#### Moving a container to a new location:

```powershell
> docker container mv myContainer E:\DockerData\nano\
```
(The corresponding container ought to be stopped or paused for this operation to succeed.)
<br/>

#### Moving a volume to a new location:

```powershell
> docker volume mv myVolume E:\DockerData\nano\
```
(All referencing containers ought to be stopped or paused (perhaps automatically temprarily be paused when accessing the volume while this operation is in progress) for this operation to succeed.)

**Comment by SetTrend** [comment]
Out of curiosity: Will I need to keep this issue alive for keeping it from auto-closing?

**Comment by stephen-turner** [comment]
We talked to the WSL engineers. They recommended this recipe: https://craigloewen-msft.github.io/WSLTipsAndTricks/tip/install-different-drive.html.

I'd be interested to know whether this works for you, or if not what the blockers are.

**Comment by SetTrend** [comment]
Yes, I'm aware of this workaround. Yet, let me share with you the following drawbacks:

- You cannot move a particular Docker image/container/volume to another drive. It's an all or nothing process. You you're not able to distribute the load across your drives.
- It doesn't work for Windows Docker images/containers/volumes.

**Comment by stephen-turner** [comment]
You're hoping for a different path for different images? I don't think we're ever going to be able to provide that if WSL 2 doesn't provide it.

**Comment by SetTrend** [comment]
Your right. Exactly.

That's what my WSL 2 proposal ([mentioned above](/microsoft/WSL/issues/7128#issuecomment-870465528)) is about:

> What I'm suggesting is to be able to store individual images, containers or volumes on **_different_** drives _**individually**_.

**Comment by stephen-turner** [comment]
OK, that's a much bigger change than I imagined. We might be able to provide help with moving everything to a single other location, but I don't see us having a location per image.
